npm/networkPolicyController.go

```diff
@@ -0,0 +1,491 @@
+	// since netPolController did not apply for any states for the network policy
+	if !netPolExists {
+		return
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	log.Logf("Starting Network Policy %d worker(s)", threadiness)
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	log.Logf("Started Network Policy %d worker(s)", threadiness)
+	<-stopCh
+	log.Logf("Shutting down Network Policy workers")
+
+	return nil
+}
+
+func (c *networkPolicyController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *networkPolicyController) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncNetPol, passing it the namespace/name string of the
+		// network policy resource to be synced.
+		// TODO : may consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type later
+		if err := c.syncNetPol(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		log.Logf("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		metrics.SendErrorLogAndMetric(util.NetpolID, "syncNetPol error due to %v", err)
+		return true
+	}
+
+	return true
+}
+
+// syncNetPol compares the actual state with the desired, and attempts to converge the two.
+func (c *networkPolicyController) syncNetPol(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	log.Logf("SyncNetPol %s %s %s", key, namespace, name)
+
+	// Get the network policy resource with this namespace/name
+	netPolObj, err := c.netPolLister.NetworkPolicies(namespace).Get(name)
+
+	// (TODO): Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Logf("Network Policy %s is not found, may be it is deleted", key)
```

Comment:
Even in these log messages we should have the [<funcName>] that way it will be easier while reading logs

Reply:
Good point. By the way why don't we use klog? It will save our time and more consistent since file names and line comes as free. Are there special reasons?

Reply:
this log lib is generic one we use for other services in this repo. If it makes sense we can move, or add such support to acn/log so our other services can reap the benefits

Reply:
Concensus opinion is ok to use klog. Replaced log with klog in networkPolicyController, which adds extra information to help reading logs.